dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java

```diff
@@ -260,55 +263,146 @@ JarLinkageReport generateLinkageReport(Path jarPath, SymbolReferenceSet symbolRe
+      }
+
+      // Checks the target class, its parent classes, and its interfaces.
+      // Interface check is needed to avoid false positive for a method reference to an abstract
+      // class that implements an interface. For example, Guava's ImmutableList is an abstract class
+      // that implements List interface, but the class does not have get() method. A method
+      // reference to ImmutableList.get() should not be reported as a linkage error.
+      Iterable<JavaClass> typesToCheck =
+          Iterables.concat(
+              getClassAndSuperClasses(targetJavaClass),
+              Arrays.asList(targetJavaClass.getAllInterfaces()));
+      for (JavaClass javaClass : typesToCheck) {
+        for (Method method : javaClass.getMethods()) {
+          if (method.getName().equals(methodName)
+              && method.getSignature().equals(reference.getDescriptor())) {
+            if (!isMemberAccessibleFrom(javaClass, method, reference.getSourceClassName())) {
+              return Optional.of(
+                  StaticLinkageError.errorInaccessibleMember(reference, classFileLocation));
+            }
+            return Optional.empty();
+          }
+        }
+      }
+
+      // The class is in class path but the symbol is not found
+      return Optional.of(StaticLinkageError.errorMissingMember(reference, classFileLocation));
+    } catch (ClassNotFoundException ex) {
+      return Optional.of(StaticLinkageError.errorMissingTargetClass(reference));
+    }
   }
 
   /**
    * Returns an {@code Optional} describing the linkage error for the field reference if the
    * reference does not have a valid referent in the input class path; otherwise an empty {@code
    * Optional}.
    */
-  private Optional<StaticLinkageError<FieldSymbolReference>> checkLinkageErrorMissingFieldAt(
+  @VisibleForTesting
+  Optional<StaticLinkageError<FieldSymbolReference>> checkLinkageErrorMissingFieldAt(
       FieldSymbolReference reference) {
     String targetClassName = reference.getTargetClassName();
     String fieldName = reference.getFieldName();
-    for (JavaClass javaClass : getClassAndSuperClasses(targetClassName)) {
-      for (Field field : javaClass.getFields()) {
-        if (field.getName().equals(fieldName)) {
-          // The field is found. Returning no error.
-          return Optional.empty();
+    try {
+      JavaClass targetJavaClass = classDumper.loadJavaClass(targetClassName);
+      Path classFileLocation = classDumper.findClassLocation(targetClassName);
+      if (!isClassAccessibleFrom(targetJavaClass, reference.getSourceClassName())) {
+        return Optional.of(StaticLinkageError.errorInaccessibleClass(reference, classFileLocation));
+      }
+
+      for (JavaClass javaClass : getClassAndSuperClasses(targetJavaClass)) {
+        for (Field field : javaClass.getFields()) {
+          if (field.getName().equals(fieldName)) {
+            // The field is found. Returning no error.
```

Comment:
You could still return a different error here though, right?

Reply:
Right. The comment was outdated. Moved to appropriate line.